### PR TITLE
RL: Compute logprobs from vocab-parallel logits

### PR DESCRIPTION
**Pull request by tdene**
- [x] I, the PR author, have personally reviewed every line of this PR.

# What does this PR do?

:warning: For major changes (either in lines of code or in its impact), please make sure to first share a design doc with the team. If you're unsure what's the best way to do so, contact @NVIDIA/mcore-oncall.

## Issue tracking

For PRs from open-source community contributors:

- **New features**: a linked issue is **required**. Please open a [feature request](https://github.com/NVIDIA/Megatron-LM/issues/new?template=feature_request.md) and reference it here before submitting the PR.
- **Small updates (bug fixes, minor improvements)**: a linked issue is **recommended** and will accelerate the PR review process.

Linked issue: 

### Pre-checks

- [ ] I have added relevant unit tests
- [ ] I have added relevant functional tests
- [ ] If this PR adds or changes a GPU kernel (Triton, `jit_fuser`/`torch.compile`, CUDA extension, TE or external-library dispatch, or a scatter/index accumulation), I have added or updated its bit-exact determinism test and registered it in `tests/unit_tests/determinism/kernels/manifest.py` ([guide](https://github.com/NVIDIA/Megatron-LM/blob/main/docs/developer/determinism/testing.md))
- [ ] I have added proper typing to my code [Typing guidelines](https://docs.python.org/3/library/typing.html)
- [ ] I have added relevant documentation
- [ ] I have run the [autoformatter.sh](https://github.com/NVIDIA/Megatron-LM/blob/main/tools/autoformat.sh) on my PR

### Code review

Feel free to message or comment @NVIDIA/mcore-oncall to help accelerate your merge into main. The less complex your PR is, the faster it will be approved and merged!

All PRs start as **draft**. If you open a non-draft PR, it will be automatically converted to draft.

#### Step 1: Mark PR as "Ready for Review"

1. When your PR is ready, click **Ready for Review**.
2. An oncall reviewer is auto-assigned and expert reviewers are notified based on your changes.
   - Some PRs may jump straight to step 2. This is determined by `.github/CODEOWNERS`.

:warning: Only mark as ready once merge-conflicts are resolved and the CI is passing.
Final Review might get declined if these requirements are not fulfilled.

#### Step 2: Final Review

For PRs that change `megatron/core`, once all expert reviewers have approved, the `Final Review` label is applied **automatically** and final reviewers are assigned.

For PRs outside `megatron/core`, this step is skipped.

#### Step 3: Approved

Once all required reviewers have approved, the `Approved` label is applied **automatically**.

### Merge

Any member of [mcore-engineers](https://github.com/orgs/NVIDIA/teams/mcore-engineers) will be able to merge your PR.

### CODE REVIEW

File: tests/unit_tests/rl/test_rl_utils.py
Comment (jalbericiola) [inline tests/unit_tests/rl/test_rl_utils.py:54]: must be TP-divisible, e.g. 768

Comment (jalbericiola) [inline tests/unit_tests/rl/test_rl_utils.py:69]: this function should be something like:
def __call__(self, x, position_ids, attention_mask, runtime_gather_output=True, **kwargs):
        del position_ids
        del attention_mask
        batch, seq = x.shape
        tp_size = get_pg_size(self.pg_collection.tp)
        width = self.vocab if runtime_gather_output else self.vocab // tp_size
        mock_model_outputs = torch.ones((batch, seq, width), device=x.device, dtype=self.dtype)
        return mock_model_outputs

Comment (jalbericiola) [inline tests/unit_tests/rl/test_rl_utils.py:51]: from megatron.core.utils import get_pg_size
import math

**Review by jalbericiola — CHANGES_REQUESTED** [review]
The comments in the test are to make it to consider sharding.

Then we need MTP loss decoupling in megatron/core/transformer/multi_token_prediction.py:
Line 1128: pass runtime_gather_output=False to the MTP head call instead of the inherited flag.
Line 1187: pass False to the acceptance-count helper so it takes the sharded argmax branch.
PR description: add the paragraph on the TP>1 behavior change, plus a TP>1 MTP test in the RL test file.
